src/devices/sound/hc55516.cpp

```diff
@@ -21,242 +31,561 @@
+//  device_start - device-specific start
+//-------------------------------------------------
 
-inline int hc55516_device::current_clock_state()
+void hc55516_device::device_start()
 {
-	return ((uint64_t)m_update_count * clock() * 2 / SAMPLE_RATE) & 0x01;
+	start_common(0x07, RISING);
+	save_item(NAME(m_sylmask));
+	save_item(NAME(m_sylshift));
+	save_item(NAME(m_syladd));
+	save_item(NAME(m_intshift));
+	save_item(NAME(m_sylfilter));
+	save_item(NAME(m_intfilter));
+	save_item(NAME(m_agc));
+	save_item(NAME(m_buffered_fzq));
+
+	/* resolve lines */
+	m_agc_push_cb.resolve();
+	m_fzq_pull_cb.resolve();
+
+	/* variant-specific parameters */
+	m_sylmask = 0xfc0;
+	m_sylshift = 6;
+	m_syladd = 0xfc1;
+	m_intshift = 4;
 }
 
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
 
-void hc55516_device::process_digit()
+void hc55516_device::device_reset()
 {
-	double integrator = m_integrator, temp;
+	// simulate /FZ having been held for a while
+	m_sylfilter = 0x3f;
+	m_intfilter = 0;
+	m_agc = true;
+	m_buffered_fzq = true; // assuming /FZ was just released and is now high/inactive
+}
 
-	/* shift the bit into the shift register */
-	m_shiftreg = (m_shiftreg << 1) | m_digit;
+// device specific functions
 
-	/* move the estimator up or down a step based on the bit */
-	if (m_digit)
-		integrator += m_filter;
-	else
-		integrator -= m_filter;
+WRITE_LINE_MEMBER( hc55516_device::fzq_w )
+{
+	m_buffered_fzq = state;
+}
+
+READ_LINE_MEMBER( hc55516_device::agc_r )
+{
+	// prevent debugger from changing the internal state
+	if (!machine().side_effects_disabled())
+		m_stream->update(); /* bring up to date first */
+	return m_agc;
+}
 
-	/* simulate leakage */
-	integrator *= m_leak;
+void hc55516_device::process_bit(bool bit, bool clock_state)
+{
+	bool frozen =  (  ( (m_intfilter >= 0x180) && (!bit) ) ||
+		( (m_intfilter <= -0x180) && (bit) )  );
 
-	/* if we got all 0's or all 1's in the last n bits, bump the step up */
-	if (((m_shiftreg & m_shiftreg_mask) == 0) ||
-		((m_shiftreg & m_shiftreg_mask) == m_shiftreg_mask))
+	int32_t sum;
+	if (is_active_clock_transition(clock_state))
 	{
-		m_filter = FILTER_MAX - ((FILTER_MAX - m_filter) * m_charge);
+		// grab the /FZ state; if the callback is present, use that, otherwise use the buffered state
+		bool fzq_state = false;
+		if (!m_fzq_pull_cb.isnull())
+			fzq_state = m_fzq_pull_cb();
+		else
+			fzq_state = m_buffered_fzq;
 
-		if (m_filter > FILTER_MAX)
-			m_filter = FILTER_MAX;
-	}
+		if (!fzq_state) // /FZ is active low, if it is active, the input bit is ignored and the inverse of the previous bit in the shifter is used instead
+			bit = !(m_shiftreg&1);
 
-	/* simulate decay */
-	else
+		/* shift the new bit into the shift register */
+		m_shiftreg = (m_shiftreg << 1) | (bit?1:0);
+
+		if ( ((m_shiftreg & m_shiftreg_mask) == 0) ||
+		((m_shiftreg & m_shiftreg_mask) == m_shiftreg_mask) )
```

Comment:
Please indent line continuations, preferably by two tabs so they’re clearly differentiated from an increase in scope level.

Reply:
Fixed, this was actually mangled during a copy-paste from the old hc55516 code, I've restored it to the same indenting as that code had (the old code still exists later in the file on lines 551-553)